File: activity_log.py
"""
Shared append-only activity logger for the admin-only Activiteitenlog in
the app - written to by sync_and_cleanup_freeshow.py and
tuya_http_server.py, read by src/lib/activityLog.ts (same file, same JSON
Lines format) via the app's /api/activity-log route.

Trimmed automatically once it grows past a size threshold, same policy as
the TypeScript side - keeps disk usage bounded without OS-level log
rotation. A rewrite landing at the same moment as a write from the Next.js
side is a low-probability, low-stakes race for what's just an admin
convenience log, so this doesn't bother with file locking either.
"""
import json
import os
import time

TRIM_THRESHOLD_BYTES = 2 * 1024 * 1024
MAX_ENTRIES = 5000


def _log_path():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    candidates = [
        os.path.join(script_dir, "data", "activity_log.jsonl"),
        "/app/data/activity_log.jsonl",
    ]
    for c in candidates:
        if os.path.isdir(os.path.dirname(c)):
            return c
    return candidates[0]


def _safe_chmod(path):
    # World-writable, matching sync_and_cleanup_freeshow.py's own
    # _safe_chmod - this file is written from both this container
    # (tuya-control) and the main app container, under different users, so
    # whichever side creates it first must leave it open for the other.
    # Best effort only: never let a chmod failure break the actual log write.
    try:
        os.chmod(path, 0o666)
    except OSError:
        pass


def _trim_if_needed(path):
    try:
        if os.path.getsize(path) <= TRIM_THRESHOLD_BYTES:
            return
        with open(path, "r", encoding="utf-8") as f:
            lines = [l for l in f.read().split("\n") if l]
        trimmed = lines[-MAX_ENTRIES:]
        with open(path, "w", encoding="utf-8") as f:
            f.write("\n".join(trimmed) + "\n")
        _safe_chmod(path)
    except FileNotFoundError:
        pass
    except Exception as e:
        print(f"[ActivityLog] Kon logboek niet inkorten: {e}")


def log_activity(category, message, details=None):
    """category: 'sync' | 'plug' | 'led' | 'error' | 'settings' | 'system'"""
    try:
        path = _log_path()
        os.makedirs(os.path.dirname(path), exist_ok=True)
        entry = {
            "ts": time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime()) + f".{int(time.time() * 1000) % 1000:03d}Z",
            "category": category,
            "message": message,
        }
        if details:
            entry["details"] = details
        with open(path, "a", encoding="utf-8") as f:
            f.write(json.dumps(entry) + "\n")
        _safe_chmod(path)
        _trim_if_needed(path)
    except Exception as e:
        print(f"[ActivityLog] Kon gebeurtenis niet wegschrijven: {e}")

File: src/lib/activityLog.ts
import fs from 'fs';
import path from 'path';

const DATA_DIR = path.join(process.cwd(), 'data');
const LOG_FILE = path.join(DATA_DIR, 'activity_log.jsonl');

// Trimmed whenever the file grows past this size, down to MAX_ENTRIES most
// recent lines - keeps disk usage bounded without needing OS-level log
// rotation (sync_cleanup.log has no such cap and has grown past 2MB).
const TRIM_THRESHOLD_BYTES = 2 * 1024 * 1024;
const MAX_ENTRIES = 5000;

export type ActivityCategory = 'sync' | 'plug' | 'led' | 'error' | 'settings' | 'system';

export interface ActivityEntry {
  ts: string;
  category: ActivityCategory;
  message: string;
  details?: Record<string, unknown>;
}

function ensureDataDir() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

// World-writable, matching sync_and_cleanup_freeshow.py's own
// _safe_chmod - this file is written from both this container and the
// separate tuya-control container (a different user), so whichever side
// creates it first must leave it open for the other. Without this, the
// Python side hits a hard "Permission denied" the moment it tries to
// append to a file this container already created with default (owner-only)
// permissions. Best effort only: never let a chmod failure break the
// actual log write.
function safeChmod(target: string) {
  try {
    fs.chmodSync(target, 0o666);
  } catch {
    // ignore
  }
}

// Rewriting the whole file is only done occasionally (once the file passes
// the size threshold), so a concurrent append from the Python side landing
// mid-rewrite is a low-probability, low-stakes race for what's just an
// admin convenience log - not worth a locking mechanism over.
function trimIfNeeded() {
  try {
    const stats = fs.statSync(LOG_FILE);
    if (stats.size <= TRIM_THRESHOLD_BYTES) return;
    const lines = fs.readFileSync(LOG_FILE, 'utf-8').split('\n').filter(Boolean);
    const trimmed = lines.slice(-MAX_ENTRIES);
    fs.writeFileSync(LOG_FILE, trimmed.join('\n') + '\n');
    safeChmod(LOG_FILE);
  } catch {
    // File doesn't exist yet, or a transient read/write error - next write
    // will just create/append normally.
  }
}

export function logActivity(category: ActivityCategory, message: string, details?: Record<string, unknown>) {
  try {
    ensureDataDir();
    const entry: ActivityEntry = { ts: new Date().toISOString(), category, message, ...(details ? { details } : {}) };
    fs.appendFileSync(LOG_FILE, JSON.stringify(entry) + '\n');
    safeChmod(LOG_FILE);
    trimIfNeeded();
  } catch (e) {
    console.error('[ActivityLog] Kon gebeurtenis niet wegschrijven:', e);
  }
}

// Newest first. `before` (an ISO timestamp) paginates further back in time.
export function readActivity(opts: { limit?: number; category?: ActivityCategory; before?: string } = {}): ActivityEntry[] {
  const limit = opts.limit ?? 100;
  try {
    if (!fs.existsSync(LOG_FILE)) return [];
    const lines = fs.readFileSync(LOG_FILE, 'utf-8').split('\n').filter(Boolean);
    const entries: ActivityEntry[] = [];
    for (const line of lines) {
      try {
        entries.push(JSON.parse(line));
      } catch {
        // Skip a corrupted line rather than failing the whole read.
      }
    }
    entries.reverse();
    let filtered = entries;
    if (opts.category) filtered = filtered.filter(e => e.category === opts.category);
    if (opts.before) filtered = filtered.filter(e => e.ts < opts.before!);
    return filtered.slice(0, limit);
  } catch (e) {
    console.error('[ActivityLog] Kon logboek niet lezen:', e);
    return [];
  }
}
